GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs

```diff
@@ -0,0 +1,169 @@
+            cacheContext.NoCache = true;
+            IList<IPackageSearchMetadata> queryResults = (await packageMetadataResource.GetMetadataAsync(packageId, true, true, cacheContext, new Logger(this.tracer), CancellationToken.None)).ToList();
+            return queryResults;
+        }
+
+        /// <summary>
+        /// Download the specified packageId from the NuGet feed.
+        /// </summary>
+        /// <param name="packageId">PackageIdentity to download.</param>
+        /// <returns>Path to the downloaded package.</returns>
+        public virtual async Task<string> DownloadPackage(PackageIdentity packageId)
+        {
+            SourceRepository sourceRepository = Repository.Factory.GetCoreV3(this.feedUrl);
+            if (!string.IsNullOrEmpty(this.personalAccessToken))
+            {
+                sourceRepository.PackageSource.Credentials = this.Credentials;
+            }
+
+            var downloadResource = await sourceRepository.GetResourceAsync<DownloadResource>();
+            var downloadResourceResult = await downloadResource.GetDownloadResourceResultAsync(packageId, new PackageDownloadContext(new SourceCacheContext(), this.downloadFolder, true), string.Empty, new Logger(this.tracer), CancellationToken.None);
+
+            string downloadPath = Path.Combine(this.downloadFolder, $"{this.feedName}.zip");
+
+            using (var fileStream = File.Create(downloadPath))
+            {
+                downloadResourceResult.PackageStream.CopyTo(fileStream);
+            }
+
+            return downloadPath;
+        }
+
+        /// <summary>
+        /// Implementation of logger used by NuGet library. It takes all output
+        /// and redirects it to the GVFS logger.
+        /// </summary>
+        public class Logger : ILogger
+        {
+            private ITracer tracer;
+
+            public Logger(ITracer tracer)
+            {
+                this.tracer = tracer;
+            }
+
+            public void Log(LogLevel level, string data)
+            {
+                this.tracer.RelatedInfo($"NuGetPackageUpgrader ({level}): {data}");
+            }
+
+            public void Log(ILogMessage message)
+            {
+                this.tracer.RelatedInfo($"NuGetPackageUpgrader ({message.Level}): {message.Message}");
+            }
+
+            public Task LogAsync(LogLevel level, string data)
+            {
+                this.Log(level, data);
+                return Task.CompletedTask;
+            }
+
+            public Task LogAsync(ILogMessage message)
+            {
+                this.Log(message);
+                return Task.CompletedTask;
+            }
+
+            public void LogDebug(string data)
+            {
+                this.tracer.RelatedInfo($"NuGetPackageUpgrader (Debug): {data}");
+            }
+
+            public void LogError(string data)
+            {
+                this.tracer.RelatedInfo($"NuGetPackageUpgrader (Error): {data}");

```

Comment:
Why are all of the Nuget log messages being mapped to `Info`? Should some any of them be mapped to warning or error?